src/broker/redis.rs

```diff
@@ -41,95 +46,187 @@ impl BrokerBuilder for RedisBrokerBuilder {
         // let blocking_conn = client.get_connection().unwrap();
 
-        let consume_channel = Channel{
-            client: client.clone(),
-            prefetch_count: self.config.prefetch_count,
-        };
+        let manager = client.get_tokio_connection_manager().await?;
 
-        Ok(RedisBroker{
+        let (tx, rx) = channel(1);
+        Ok(RedisBroker {
             uri: self.config.broker_url.clone(),
-            conn: Mutex::new(multiplexed_conn.clone()),
             queues: queues,
-            prefetch_count: Mutex::new(self.config.prefetch_count),
-            consume_channel: consume_channel,
-            consume_channel_write_lock: Mutex::new(0),
+            client: client,
+            manager: manager,
+            prefetch_count: Arc::new(AtomicU16::new(self.config.prefetch_count)),
+            pending_tasks: Arc::new(AtomicU16::new(0)),
+            waker_rx: Mutex::new(rx),
+            waker_tx: tx,
         })
     }
 }
 
-pub struct RedisBroker{
+pub struct RedisBroker {
     uri: String,
     /// Broker connection.
-    conn: Mutex<MultiplexedConnection>,
-    consume_channel: Channel,
+    client: Client,
+    manager: ConnectionManager,
     /// Mapping of queue name to Queue struct.
-    consume_channel_write_lock: Mutex<u8>,
     queues: HashSet<String>,
 
     /// Need to keep track of prefetch count. We put this behind a mutex to get interior
     /// mutability.
-    prefetch_count: Mutex<u16>,
+    prefetch_count: Arc<AtomicU16>,
+    pending_tasks: Arc<AtomicU16>,
+    waker_rx: Mutex<Receiver<Waker>>,
+    waker_tx: Sender<Waker>,
 }
 
-pub struct Channel{
-    client: Client,
-    prefetch_count: u16,
+#[derive(Clone)]
+pub struct Channel {
+    connection: ConnectionManager,
+    queue_name: String,
 }
 
-impl fmt::Debug for Channel{
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result{
-        write!(f, "Channel {{ client: MultiplexedConnection }}")
+impl fmt::Debug for Channel {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Channel {{ {} }}", self.queue_name)
+    }
+}
+
+impl Channel {
+    fn new(connection: ConnectionManager, queue_name: String) -> Self {
+        Self {
+            connection,
+            queue_name,
+        }
+    }
+
+    fn pending_queue(&self) -> String {
+        format!("_celery.{}_pending", self.queue_name)
+    }
+
+    fn process_map_name(&self) -> String {
+        format!("_celery.{}_process_map", self.queue_name)
+    }
+
+    async fn fetch_task(
+        mut self,
+        send_waker: Option<(Sender<Waker>, Waker)>,
+    ) -> Result<Delivery, BrokerError> {
+        let pending_queue = self.pending_queue();
+        if let Some((mut sender, waker)) = send_waker {
+            sender.send(waker).await.unwrap();
+            futures::pending!();
+        }
+        loop {
+            let rez: Result<Option<String>, RedisError> = redis::cmd("RPOPLPUSH")
```

Comment:
We just need to do RPOP here.

Reply:
- - - - - - - 
- __



- Pruning
- - - 












- 










- -